vignettes/articles/oauth.Rmd

```diff
@@ -13,39 +13,58 @@ knitr::opts_chunk$set(
-The process varies from API to API, but at the end of it you'll get a client id and in most cases a client secret.
+The first step in working with any OAuth API is to create an **application** **client**, which I'll generally just call a **client**.
+You'll do this by registering for a developer account on the API's website then creating a new OAuth app.
+This process varies from API to API (it's normal to spend some time hunting through the docs and settings), but at the end of it you'll get a client id (another random string of numbers and letters).
 
-(You'll definitely need this for testing your package, and you'll probably also baked it into your package for the convenience of your users. Bundling the app is user friendly, but not always possible, particularly if rate limits are enforced on a per-app rather than per-user basis. You should always provide some way for the user to provide their own app.)
+Sometimes the client id is all you need, and you can give httr2 the information about the client with a call to `oauth_client()`, something like this:
 
-If the API provides a way to authenticate your app without the client secret, you should leave it out of your package.
-But in most cases, you'll need to include the secret in the package.
-You can use `obfuscate()` to hide the secret; this is not bulletproof[^2], but in most cases it'll be easier to create a new client than try and steal yours.
-Additionally, it's unusual for an OAuth client to be able to do anything in its own right, so even if someone does steal your secret there's not much harm they can do with it.
+```{r}
+client <- oauth_client(
+  id = "28acfec0674bb3da9f38",
+  token_url = "https://github.com/login/oauth/access_token",
+  name = "hadley-oauth-test-2"
+)
+```
+
+The call to `oauth_client()` also includes a `name` and a `token_url`.
+The `name` is human facing and should typically be the same as the package, if you're using one.
+I have a bunch of apps that I've used for testing, so here I've and is used to remind you which app the client corresponds to.
+The `token_url` points to the URL that's used to obtain an access token.
+You'll need to find this from the documentation for the API you're wrapping; it will typically be found in the section that describes the OAuth process and will be an endpoint that returns an access token.
 
-[^2]: It uses `secret_encrypt()` with a special encryption key that's bundled with httr2.
+### Client secret
 
-To obfuscate a secret, call `obfuscate()`:
+In most cases, however, the API will also require a client secret.
+While this is called a secret, it's typically not that important to keep it a secret because of two reasons:
+
+-   It's typically easy to create an new app on the developer website so stealing yours wouldn't save much time.
+-   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.
```

Comment:
Yeah but there’s no reason they can’t create their own app that has the same details as your app.

Reply:
Hmm. Good point.

Reply:
I might add a sentence saying something like "as long as you didn't pay to get the app, or have to supply any secret information when creating it, there's no reason to worry about it"

Reply:
I also feel like the word "secret" is a huge reason people assume that this needs to be kept secret in the same way that a password does. And it is a reasonable assumption!

Reply:
The case I can think of where the secret matters is something like a Slack app, where the app (which can be pretty much just an API client) gets approved for a given Slack workspace, and thus you can do a lot if you successfully get someone to log into the app. That's assuaged online by restricting redirect urls, but it's messy if you're hitting the API from your machine (where the redirect is going to be localhost).

That's a case where it might be best to not supply the client, although I'm still unsure how to feel about it. I guess a way to protect it is to make it clear what the app is supposed to be paired with (if you're able to do so on the approval page), but it's an area that always feels scary.